### PR TITLE
Use prefixing MD5 hasher by default

### PR DESCRIPTION
**Pull request by lptr**
We have been using both a "raw" MD5 hasher (where we simply feed the data to the hasher) and a prefixing MD5 hasher (where we also prefix each hashed data unit with the length of the data to avoid collisions between hashing `aabb` vs `aa` + `bb`).

This PR makes the prefixing hasher the default. It also makes MD5 the default. This also makes it easier to later switch the internal hashing algorithm to something else if we want to.

### CODE REVIEW

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=15583871

File: subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java [hunk 1]
```diff
@@ -196,4 +278,91 @@ public void putHash(HashCode hashCode) {
             putBytes(hashCode.getBytes());
         }
     }
+
+    @VisibleForTesting
```
Comment (wolfs) [inline subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java]: `@VisibleForTesting` is not necessary any more.

File: subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java [hunk 2]
```diff
@@ -196,4 +278,91 @@ public void putHash(HashCode hashCode) {
...
+
+        @Override
+        public boolean isValid() {
+            return valid;
+        }
+
+        @Override
+        public HashCode hash() {
+            if (!valid) {
+                throw new IllegalStateException("Build cache hash is not valid");
```
Comment (wolfs) [inline subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java]: I guess the exception should be `Hash is not valid`.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/AbiExtractingClasspathResourceHasher.java [hunk 3]
```diff
@@ -47,7 +47,9 @@ private HashCode hashClassBytes(InputStream inputStream) throws IOException {
         if (extractor.shouldExtractApiClassFrom(reader)) {
             byte[] signature = extractor.extractApiClassFrom(reader);
             if (signature != null) {
-                return Hashing.md5().hashBytes(signature);
+                Hasher hasher = Hashing.newHasher();
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/AbiExtractingClasspathResourceHasher.java]: Why don't we use `Hashing.hashBytes` here? I don't see any reason to use `newHasher`.

File: subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java (hunk 1 above)
Comment (lptr) [reply subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java]: Good catch, made it private.

File: subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java (hunk 2 above)
Comment (lptr) [reply subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java]: Yep.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/AbiExtractingClasspathResourceHasher.java (hunk 3 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/AbiExtractingClasspathResourceHasher.java]: Changed.

**Pull request merged**